src/document-processor.ts

```diff
@@ -0,0 +1,400 @@
+ * HTML -> HTML conversion.
+ */
+export abstract class DocumentProcessor {
+  protected readonly originalUrl: OriginalDocumentUrl;
+  protected readonly convertedUrl: ConvertedDocumentUrl;
+  protected readonly urlHandler: UrlHandler;
+  protected readonly conversionSettings: ConversionSettings;
+  protected readonly document: Document;
+  protected readonly program: Program;
+  protected readonly convertedHtmlScripts: ReadonlySet<Import>;
+
+  constructor(
+      document: Document, urlHandler: UrlHandler,
+      conversionSettings: ConversionSettings) {
+    this.conversionSettings = conversionSettings;
+    this.urlHandler = urlHandler;
+    this.document = document;
+    this.originalUrl = urlHandler.getDocumentUrl(document);
+    this.convertedUrl = this.convertDocumentUrl(this.originalUrl);
+    ({program: this.program, convertedHtmlScripts: this.convertedHtmlScripts} =
+         this.prepareJsModule());
+  }
+
+  private isInternalNonModuleImport(scriptImport: Import): boolean {
+    const oldScriptUrl = this.urlHandler.getDocumentUrl(scriptImport.document);
+    const newScriptUrl = this.convertScriptUrl(oldScriptUrl);
+    const isModuleImport =
+        dom5.getAttribute(scriptImport.astNode, 'type') === 'module';
+    const isInternalImport =
+        this.urlHandler.isImportInternal(this.convertedUrl, newScriptUrl);
+    return isInternalImport && !isModuleImport;
+  }
+
+  /**
+   * Creates a single program from all the JavaScript in the current document.
+   * The standard program result can be used for either scanning or conversion.
+   */
+  private prepareJsModule() {
+    const combinedToplevelStatements = [];
+    const convertedHtmlScripts = new Set<Import>();
+    const claimedDomModules = new Set<parse5.ASTNode>();
+    let prevScriptNode: parse5.ASTNode|undefined = undefined;
+    for (const script of this.document.getFeatures()) {
+      let scriptDocument: Document;
+      if (script.kinds.has('html-script') &&
+          this.isInternalNonModuleImport(script as Import)) {
+        scriptDocument = (script as Import).document;
+        convertedHtmlScripts.add(script as Import);
+      } else if (script.kinds.has('js-document')) {
+        scriptDocument = script as Document;
+      } else {
+        continue;
+      }
+      const scriptProgram =
+          recast.parse(scriptDocument.parsedDocument.contents).program;
+      rewriteToplevelThis(scriptProgram);
+      removeToplevelUseStrict(scriptProgram);
+      // We need to inline templates on a per-script basis, otherwise we run
+      // into trouble matching up analyzer AST nodes with our own.
+      const localClaimedDomModules =
+          this.inlineTemplates(scriptProgram, scriptDocument);
+      for (const claimedDomModule of localClaimedDomModules) {
+        claimedDomModules.add(claimedDomModule);
+      }
+      if (this.conversionSettings.addImportPath) {
+        this.addImportPathsToElements(scriptProgram, scriptDocument);
```

Comment:
This feels like a "conversion" concern, can it be moved to DocumentConverter?

Reply:
I don't want to move this mutation out of here yet, my goal here was for this to be a simple noop code split. Added a comment+TODO